command/agent/auth/oci/oci.go

```diff
@@ -0,0 +1,262 @@
+	lastCreds string
+
+	// Notifies the outer environment that it should call Authenticate again.
+	credsFound chan struct{}
+
+	// Detects that the outer environment is closing.
+	stopCh chan struct{}
+}
+
+func (a *ociMethod) Authenticate(context.Context, *api.Client) (string, http.Header, map[string]interface{}, error) {
+	a.credLock.Lock()
+	defer a.credLock.Unlock()
+
+	a.logger.Trace("beginning authentication")
+
+	requestPath := fmt.Sprintf("/v1/%s/login/%s", a.mountPath, a.role)
+	requestURL := fmt.Sprintf("%s%s", a.vaultAddress, requestPath)
+
+	request, err := http.NewRequest("GET", requestURL, nil)
+	if err != nil {
+		return "", nil, nil, fmt.Errorf("error creating authentication request: %w", err)
+	}
+
+	// request.Header.Set("content-type", "application/json")
+	request.Header.Set("Date", time.Now().UTC().Format(http.TimeFormat))
+
+	signer := common.DefaultRequestSigner(a.configurationProvider)
+
+	err = signer.Sign(request)
+
+	if err != nil {
+		return "", nil, nil, fmt.Errorf("error signing authentication request: %w", err)
+	}
+
+	parsedVaultAddress, err := url.Parse(a.vaultAddress)
+	if err != nil {
+		return "", nil, nil, fmt.Errorf("unable to parse vault address: %w", err)
+	}
+
+	request.Header.Set("Host", parsedVaultAddress.Host)
+	request.Header.Set("(request-target)", fmt.Sprintf("%s %s", "get", requestPath))
+
+	data := map[string]interface{}{
+		"request_headers": request.Header,
+	}
+
+	return fmt.Sprintf("%s/login/%s", a.mountPath, a.role), nil, data, nil
+}
+
+func (a *ociMethod) NewCreds() chan struct{} {
+	return a.credsFound
+}
+
+func (a *ociMethod) CredSuccess() {}
+
+func (a *ociMethod) Shutdown() {
+	close(a.credsFound)
+	close(a.stopCh)
+}
+
+func (a *ociMethod) pollForCreds(frequencySeconds int) {
+	ticker := time.NewTicker(time.Duration(frequencySeconds) * time.Second)
+	defer ticker.Stop()
+	for {
+		select {
+		case <-a.stopCh:
+			a.logger.Trace("shutdown triggered, stopping OCI auth handler")
+			return
+		case <-ticker.C:
+			if err := a.checkCreds(); err != nil {
+				a.logger.Warn("unable to retrieve current creds, retaining last creds", "error", err)
+			}
+		}
+	}
+}
+
+func (a *ociMethod) checkCreds() error {
+	a.credLock.Lock()
+	defer a.credLock.Unlock()
+
+	a.logger.Trace("checking for new credentials")
+	currentCreds, err := a.configurationProvider.KeyID()
+	if err != nil {
+		return err
+	}
+	// These will always have different pointers regardless of whether their
+	// values are identical, hence the use of DeepEqual.
+	if currentCreds == a.lastCreds {
+		a.logger.Trace("credentials are unchanged")
+		return nil
+	}
+	a.lastCreds = currentCreds
+	a.logger.Trace("new credentials detected, triggering Authenticate")
+	a.credsFound <- struct{}{}
+	return nil
+}
+
+func getHomeFolder() string {
+	current, e := user.Current()
+	if e != nil {
+		// Give up and try to return something sensible
+		home := os.Getenv("HOME")
+		if home == "" {
+			home = os.Getenv("USERPROFILE")
+		}
+		return home
```

Comment:
There's a chance that this is `""` too, right? I think I'd prefer this function also returning an error, and returning `"", e` here if we can't get a home folder.

Reply:
These were copied from `oci-go-sdk` from:
- https://github.com/oracle/oci-go-sdk/blob/87b9ee7d7a452d7d5ac6a117b314fdc74eb4ca95/common/client.go#L304
- https://github.com/oracle/oci-go-sdk/blob/87b9ee7d7a452d7d5ac6a117b314fdc74eb4ca95/common/client.go#L304

The reason I had to copy them was because they were private and the implementation of `DefaultConfigProvider` in https://github.com/oracle/oci-go-sdk/blob/87b9ee7d7a452d7d5ac6a117b314fdc74eb4ca95/common/client.go#L323 looks for configuration in `$HOME/.oci/config`, `$HOME/.obmcs/config` and environment variables prefixed with `TF_VAR_` in that order, which is not consistent with Vault's convention of environment variables taking the most precident. 

We can check if the returned value is `""`, here: https://github.com/hashicorp/vault/pull/19260/files#diff-9ba3c0a0b8a7e9e49c8d5d26e330163a5fdf8cc19a3a3adad51a87b82230e54aR96 and return an error and not proceed, but the implementation was copied from https://github.com/oracle/oci-go-sdk/blob/87b9ee7d7a452d7d5ac6a117b314fdc74eb4ca95/common/client.go#L323, with adjustments to the order we retrieve the configuration.

Reply:
Now that I think abou tit, we can't really return an error if we can't find the home directory, as we need to look for configuration in  `$HOME/.oci/config`, `$HOME/.obmcs/config` and environment variables prefixed with `OCI_` using the [`ComposingConfigurationProvider`](https://pkg.go.dev/github.com/oracle/oci-go-sdk/v65@v65.32.0/common#ComposingConfigurationProvider) and the provider will actually return an error if the credentials cannot be found.

Reply:
I'm okay with sticking with the same implementation as the SDK for consistency. Thanks for the explanation!